src/engine/joinOrdering/QueryGraph.cpp

```diff
@@ -0,0 +1,412 @@
+  return r.contains(n);
+}
+template <typename N>
+requires RelationAble<N> void QueryGraph<N>::rm_relation(const N& n) {
+  // removing a relation by turning-off all it's neighbours
+  // hiding the connections instead of erasing them, effectively deleting them.
+  for (auto& [x, e] : r[n]) {
+    r[x][n].hidden = true;
+    r[n][x].hidden = true;
+  }
+}
+template <typename N>
+requires RelationAble<N>
+void QueryGraph<N>::add_rjoin(const N& a, const N& b, float s, Direction dir) {
+  // TODO: assert single parent here?
+
+  // add connection between a -> b
+  r[a][b] = RJoin(dir);
+
+  // add connection between b -> a
+  r[b][a] = RJoin(inv(dir));
+
+  // TODO: avoid overwriting selectivity
+  // selectivity is a relation property
+  switch (dir) {
+    case Direction::UNDIRECTED:
+      if (!selectivity.contains(a)) selectivity[a] = s;
+      if (!selectivity.contains(b)) selectivity[b] = s;
+      break;
+    case Direction::PARENT:
+      if (!selectivity.contains(b)) selectivity[b] = s;
+      break;
+    case Direction::CHILD:
+      if (!selectivity.contains(a)) selectivity[a] = s;
+      break;
+  }
+}
+
+template <typename N>
+requires RelationAble<N> bool QueryGraph<N>::has_rjoin(const N& a, const N& b) {
+  // does relation a exists
+  // does relation b exists
+  // is there a connection between a and b
+  // is there a connection between b and a
+  // is the connection between a and is NOT hidden
+  return (r.contains(a) && r.contains(b) && r.at(a).contains(b) &&
+          r.at(b).contains(a) && !r[a][b].hidden);
+}
+
+template <typename N>
+requires RelationAble<N> void QueryGraph<N>::rm_rjoin(const N& a, const N& b) {
+  //    r[a].erase(b);
+  //    r[b].erase(a);
+
+  // hide the connection between a and b [dir]
+  r[a][b].hidden = true;
+
+  // hide the connection between b and a [inv(dir)]
+  r[b][a].hidden = true;
+}
+
+template <typename N>
+requires RelationAble<N>
+N QueryGraph<N>::combine(const N& a,
+                         const N& b) {  // -> Compound Relation (hist)
+
+  // 104/637
+  // if the ordering violates the query constraints, it constructs compounds
+  // TODO: assert chain
+  //    std::cout << "COMBINE " << a.label << "  " << b.label << "\n";
+
+  // 118/637
+
+  // "its cardinality is computed by multiplying the cardinalities of
+  // all relations in A and B"
+  auto w = cardinality[a] * cardinality[b];
+
+  // "its selectivity is the product of all selectivities (s_i) of relations
+  // R_i contained in A and B"
+  auto s = selectivity[a] * selectivity[b];
+
+  // add the newly computed cardinality to the
+  // cardinality map of the query graph.
+  auto n = this->add_relation(N(a.getLabel() + "," + b.getLabel(), w));
+
+  selectivity[n] = s;  // redundant
+  cardinality[n] = w;  // redundant
+
+  //    hist[n].push_back(a);
+  //    hist[n].push_back(b);
+
+  // to be able to apply the inverse operation (QueryGraph::uncombine)
+  // we keep track of the combined relation in the `hist` map
```

Comment:
As I understand it, the rank function of the combined nodes are wrong, because the cost function is wrong.
I think we should make a plan on how to split this up into a simpler design that is more easily testable.